mipengine/controller/algorithm_executor.py

```diff
@@ -0,0 +1,583 @@
+        for node in self._local_nodes:
+            for (variable_name, table_name) in node.initial_view_tables.items():
+                if variable_name in tmp_variable_node_table:
+                    tmp_variable_node_table[variable_name].update({node: table_name})
+                else:
+                    tmp_variable_node_table[variable_name] = {node: table_name}
+
+        self._initial_view_tables = {
+            variable_name: _LocalNodeTable(node_table)
+            for (variable_name, node_table) in tmp_variable_node_table.items()
+        }
+
+    @property
+    def initial_view_tables(self):
+        return self._initial_view_tables
+
+    @property
+    def algorithm_parameters(self):
+        return self._algorithm_parameters
+
+    @property
+    def x_variables(self):
+        return self._x_variables
+
+    @property
+    def y_variables(self):
+        return self._y_variables
+
+    # UDFs functionality
+    def run_udf_on_local_nodes(
+        self,
+        func_name: str,
+        positional_args: Dict[_LocalNodeTable],
+        share_to_global: bool = False,
+    ):  # -> GlobalNodeTable or LocalNodeTable
+        # queue exec_udf task on all local nodes
+        # wait for all nodes to complete the tasks execution
+        # one new table per local node was generated
+        # queue create_remote_table on global for each of the generated tables
+        # create merge table on global node to merge the remote tables
+
+        command_id = get_next_command_id()
+
+        tasks = {}
+        for node in self._local_nodes:
+            # TODO get the nodes from the LocalNodeTables in the positional_args
+            positional_args_transfrormed = []
+            keyword_args_transformed = {}
+            for var_name, val in positional_args.items():
+                if isinstance(val, _LocalNodeTable):
+                    udf_argument = UDFArgument(
+                        kind=UDFArgumentKind.TABLE,
+                        value=val.nodes_tables[node].full_table_name,
+                    )
+                elif isinstance(val, _GlobalNodeTable):
+                    raise Exception(
+                        f"(run_udf_on_local_nodes) GlobalNodeTable types are not "
+                        f"accepted from run_udf_on_local_nodes"
+                    )
+                else:
+                    udf_argument = UDFArgument(kind=UDFArgumentKind.LITERAL, value=val)
+                positional_args_transfrormed.append(udf_argument.json())
+                # keyword_args_transformed[var_name] = udf_argument.json()
+
+            task = node.queue_run_udf(
+                command_id=command_id,
+                func_name=func_name,
+                positional_args=positional_args_transfrormed,
+                keyword_args={},
+            )
+            tasks[node] = task
+
+        udf_result_tables = {}
+        for node, task in tasks.items():
+            table_name = _TableName(task.get())
```

Comment:
task.get() is a celery method, correct? Should it be called here outside of node tasks interface implementation?

Reply:
very good catch! I did not see that and it ll need a bit of thinking on how to deal with that..

Reply:
Though on second thought, because of python being a dynamically typed language, the interpreter will just try to call the `get()` method on the `task` object on runtime, no matter if the `task` object is a `celery.result.AsyncResult` or whatever other type of object. As long as the `task` object has a `get()` method, the method call will work (duck-typing), So saying that the `get()` method is a Celery method has no meaning in the context of a dynamically typed language. The interpreter just calls this method on an object, if the object has such a method the call works, if not, an `AttributeError` will be raised

So I believe the discrepancy is actually on the returning type of `node_tasks_handler_interface::INodeTasksHandler::queue_run_udf(..)` which should be a type that has a `get()` method and not a `celery.result.AsyncResult`, as it is supposedly now,
Although it is not even defined as such right now
`def queue_run_udf(...) -> "AsyncResult":`

Anyhow, I think I ll deal with that when I write the tests for AlgorithmExecutor ([MIP-237](https://team-1617704806227.atlassian.net/browse/MIP-237)) since the abstraction of the node_tasks_handler functionality was mainly implemented to allow easier implementation of dummy INodeTasksHandler objects for testing AlgorithmExecutor in isolation